text/1042-fastapi.rst

```diff
@@ -0,0 +1,357 @@
+
+
+ORM
+===
+
+* No concept of session. We're implementing simple Django-inspired API for the masses.
+
+* Objects will implement ``__eq__`` and ``__hash__`` based on Gel's object ID. We will always be fetching Gel's ID (implicitly). This mimics Django, Ruby's active record, etc.
+
+  - not yet saved objects won't hash or eq, raising an error.
+
+* Reflected Gel types will have class methods on them to build queries, e.g. ``User.select(..)``, ``User.filter(..)``, etc.
+
+* Python's ``User.select()`` will be equivalent to ``select User { * }`` in EdgeQL -- this is the only way how we can make Python limited typing work.
+
+* We will also be *always* fetching all link properties by default.
+
+* We'll be adding a new ``lazy`` annotation / field to exclude properties/computeds/link-properties from splats.
+
+* Objects can be mutated and saved: ``user.name = 'Peter'; await db.save(User(name='Anna'), user)``
+
+* For every Gel's type we'll be generating a number of Python classes:
+
+  - one for "vanilla" top-level select / insert.
+
+  - one for every link this type can be reached for -- this is needed for properly typing link properties. E.g. in the following example::
+
+    .. code-block:: python
+
+        User.select(friends=User.friends.select())
+
+    a. ``User`` will be the "vanilla" Gel's ``User`` type:
+
+       .. code-block:: python
+
+           class User:
+
+               name: str
+               friends: list[User__friends__User]
+
+    b. where ``User__friends__User`` will be a special variant of ``User`` reachable by traversing the ``User.friends`` link:
+
+       .. code-block:: python
+
+           class User__friends__User(User):
+
+               class __linkprops__:
+                   weight: int | None
+
+  - This enables link properties to always be fetched and never shadow regular properties or links. Accessing them will be possible via the ``__linkprops__`` property, e.g.:
+
+    .. code-block:: python
+
+        user = await db.get(User.select(friends=User.friends.select()))
+
+        weight = user.friends.__linkprops__.weight  # type: `int | None`
+
+* ``.id`` is a required property in Gel, however, in Python *new* objects won't have an ``.id`` until they pass through a ``save()`` call. This presents a dilemma: should the ORM type's ``id`` property be optional or required? We say it will be required, because objects will be read more often than written, so we want to optimize for the common case. Accessing the ``id`` property on an unsaved object will raise a runtime error.
+
+  - That said, the reflected types will have a custom ``__init__`` that will not accept an ``id`` argument (in disagreement with the type annotation for the field being "required").
+
+  - The custom ``__init__`` will also require passing all required links, which would be in disagreement with all links being "optional" in Python by default.
```

Comment:
We have server-side defaults for links and props. The links and props themselves might be required, but don't need to be specified because of the server-side default. This should probably use a similar mechanism to the default value for the `id`.

Reply:
Yeah, I think in general we have some ambiguity about what "required" means from the user's perspective. In Zod, for instance, we have a concept of "input" vs "output" type that let's defaulted values be optional when parsing input, while reflecting that the value is in fact always there in the parsed output. Not sure if there is some similar concept in Pydantic/Python types.